apps/admin/backend/src/cast_vote_records.ts

```diff
@@ -0,0 +1,283 @@
+ */
+function validateCastVoteRecordAgainstElectionDefinition(
+  castVoteRecord: CVR.CVR,
+  electionDefinition: ElectionDefinition
+): Result<void, CastVoteRecordElectionDefinitionValidationError> {
+  function wrapError(
+    error: Omit<CastVoteRecordElectionDefinitionValidationError, 'type'>
+  ): Result<void, CastVoteRecordElectionDefinitionValidationError> {
+    return err({ ...error, type: 'invalid-cast-vote-record' });
+  }
+
+  const { election, electionHash } = electionDefinition;
+
+  if (
+    castVoteRecord.ElectionId !== electionHash &&
+    !isFeatureFlagEnabled(
+      BooleanEnvironmentVariableName.SKIP_CVR_ELECTION_HASH_CHECK
+    )
+  ) {
+    return wrapError({ subType: 'election-mismatch' });
+  }
+
+  const precinct = getPrecinctById({
+    election,
+    precinctId: castVoteRecord.BallotStyleUnitId,
+  });
+  if (!precinct) {
+    return wrapError({ subType: 'precinct-not-found' });
+  }
+
+  const ballotStyle = getBallotStyle({
+    ballotStyleId: castVoteRecord.BallotStyleId,
+    election,
+  });
+  if (!ballotStyle) {
+    return wrapError({ subType: 'ballot-style-not-found' });
+  }
+
+  const contestValidationResult = castVoteRecordHasValidContestReferences(
+    castVoteRecord,
+    getContests({ ballotStyle, election })
+  );
+  if (contestValidationResult.isErr()) {
+    return wrapError({ subType: contestValidationResult.err() });
+  }
+
+  return ok();
+}
+
+/**
+ * Imports cast vote records given a cast vote record export directory path
+ */
+export async function importCastVoteRecords(
+  store: Store,
+  exportDirectoryPath: string
+): Promise<Result<CvrFileImportInfo, ImportCastVoteRecordsError>> {
+  const electionId = assertDefined(store.getCurrentElectionId());
+  const { electionDefinition } = assertDefined(store.getElection(electionId));
+
+  const readResult = await readCastVoteRecordExport(exportDirectoryPath);
+  if (readResult.isErr()) {
+    return readResult;
+  }
+  const { castVoteRecordExportMetadata, castVoteRecords } = readResult.ok();
+  const { castVoteRecordReportMetadata } = castVoteRecordExportMetadata;
+
+  const exportDirectoryName = path.basename(exportDirectoryPath);
+  // Hashing the export metadata, which includes a root hash of all the individual cast vote
+  // records, gives us a complete hash of the entire export
+  const exportHash = sha256(JSON.stringify(castVoteRecordExportMetadata));
+  const exportedTimestamp = castVoteRecordReportMetadata.GeneratedDate;
+
+  // Ensure that the records to be imported match the mode (test vs. official) of previously
+  // imported records
+  const mode: CvrFileMode = isTestReport(castVoteRecordReportMetadata)
+    ? 'test'
+    : 'official';
+  const currentMode = store.getCurrentCvrFileModeForElection(electionId);
+  if (currentMode !== 'unlocked' && mode !== currentMode) {
+    return err({ type: 'invalid-mode', currentMode });
+  }
+
+  const existingImportId = store.getCastVoteRecordFileByHash(
```

Comment:
Throughout the code and UI, we talk about a singular CVR file per export, which is no longer the case. I'm going to update terminology in a follow-up PR